File: tiny.py
#!/usr/bin/env python3 -B
#vim : set et sts=2 sw=2 ts=2 : 
# https://en.wikipedia.org/wiki/Effect_size#Other_metrics
from fileinput import FileInput as file_or_stdin
import traceback,random,math,sys,re
from copy import deepcopy
from termcolor import colored
from functools import cmp_to_key
from ast import literal_eval

class BOX(object):
  def __init__(i,**d): i.__dict__.update(**d)
  def __repr__(i):
    show = lambda x: x.__name__ if callable(x) else (f"{x:3g}" if isinstance(x,float) else x)
    d    = i.__dict__
    return "{"+" ".join([f":{k} {show(d[k])}" for k in d if  k[0] != "_"])+"}"

the = BOX(seed = 1234567891, 
          min  = .5,
          rest = 3,
          beam = 10,
          file = "../data/auto93.csv")
#----------------------------------------------------------------------------------------
def COL(at=0, name=""):
  return (NUM if name and name[0].isupper() else SYM)(at=at, name=name)

def SYM(at=0, name=""):
  return BOX(this=SYM, name=name, at=at, n=0, has={}, most=0, mode=None, prep=lambda x:x)

def NUM(at=0, name=""):
  return BOX(this=NUM, name=name, at=at, n=0, mu=0, m2=0, 
             hi= -1E30, lo=1E30, prep=int,
             want= 0 if (name and name[-1]) =="-" else 1)

def COLS(names):
  x,y,klass, = [], [], None
  all = [COL(at,name) for at,name in enumerate(names)]
  for col in all:
    z = col.name[-1]
    if z != "X":
      (y if z in "+-!" else x).append(col)
      if z=="!": klass=col
  return BOX(this=COLS, x=x, y=y, names=names, all=all,klass=klass)

def DATA(rows=[], data=None):
  data = data or BOX(this=DATA,rows=[],cols=None)
  [adds(data,lst) for lst in rows]
  return data
#----------------------------------------------------------------------------------------
def add(col,x):
  def _num():
    col.lo = min(x, col.lo)
    col.hi = max(x, col.hi)
    d       = x - col.mu
    col.mu += d/col.n
    col.m2 += d*(x - col.mu)
    if type(x) == float: col.prep = float
  def _sym():
    tmp = col.has[x] = 1 + col.has.get(x,0)
    if tmp > col.most: col.most, col.mode = tmp, x
  if x != "?":
    col.n += 1
    _num() if col.this is NUM  else _sym()

def norm(col,x):
  return x if (x=="?" or col.this is SYM) else (x-col.lo)/(col.hi-col.lo + 1/1E30)

def mid(col):
  return col.mode if col.this is SYM else col.mu

def div(col):
  def _ent(ns) : return -sum((n/col.n*math.log(n/col.n,2) for n in ns))
  def _std()   : return (col.m2/(col.n-1))**.5
  return _std() if col.this is NUM else _ent(col.has.values()) 

def cross(num1,num2)
  mu1, std1 = num1.mu, div(num1)
  mu2, std2 = num2.mu, div(num2)
  if mu2 < mu1: return cross(num2, num1)
  if std1==0 or std2==0 or std1==std2: return (mu1+mu2)/2
  a  = 1/(2*std1**2) - 1/(2*std2**2)
  b  = mu2/(std2**2) - mu1/(std1**2)
  c  = mu1**2 /(2*std1**2) - mu2**2 / (2*std2**2) - math.log(std2/std1)
  x1 = (-b + (b**2 - 4*a*c)**.5)/(2*a)
  x2 = (-b - (b**2 - 4*a*c)**.5)/(2*a)
  return x1 if mu1 <= x1 <= mu2 else x2
#----------------------------------------------------------------------------------------
def adds(data, row):
  def _create(): data.cols = COLS(row)
  def _update(): [add(col,row[col.at]) for col in data.cols.all]
                 data.rows += [row]
  _update() if data.cols else _create()

def d2h(data,row):
  return (sum((col.want - norm(col,row[col.at]))**2 for col in data.cols.y) / len(data.cols.y))*.5

def ordered(data,rows=None):
  return sorted(rows or data.rows, key=lambda r: d2h(data,r))

def stats(data, cols=None, fun=lambda c: mid(c)):
  return BOX(N=len(data.rows), **{col.name: fun(col) for col in (cols or data.cols.y)})

def clone(data,rows=[]):
  return DATA(data=DATA(rows=[data.cols.names]),rows=rows)
#----------------------------------------------------------------------------------------
def KEY(col,lo,hi):
  return BOX(this=KEY, at=col.at, lo=lo, hi=hi, name=col.name)

def RULE(keys=[]):
  rule = BOX(this=RULE, score = 0, cols={})
  [key(rule,k) for k in keys]
  return rule

def key(rule, new):
  if new.this is RULE:
    [key(rule, k) for keys in new.cols.values() for k in keys]
  else:
    b4 = rule.cols[new.at] = rule.cols.get(new.at,[])
    dontAdd =  False
    for  k in b4:
      if new.lo <= k.lo and k.lo <= new.hi <= k.hi: k.hi=new.hi; dontAdd=True
      if new.hi >= k.hi and k.lo <= new.lo <= k.hi: k.lo=new.lo; dontAddflag=True
    if not dontAdd: 
      b4 += [new]
  return rule

def selects(rule,row):
  def _or(keys4col):
    for key in keys4col:
      x = row[key.at]
      if x=="?" or key.lo <= x <= key.hi: return True
  for keys4col in rule.cols.values():
    if not _or(keys2col): return False
  return True

def score(rule, bests, rests):
  bs = [row for row in best.rows if selects(rule,row)]
  rs = [row for row in rest.rows if selects(rule,row)]
  b  = len(bs) / (len(bests.rows) + 1/1E30)
  r  = len(rs) / (len(rests.rows) + 1/1E30)
  rule.score = b**2/(b+r)
  yield rule

def rules(bests,rests):
  for best,rest in zip(bests.cols.x, rests.cols.x):
    for key in keys(best,rest):
      yield score(RULE([ key ]), bests,rests)

def keys(best,rest):
  "Returns ranges (lo,hi) being values more often in `best` than `rest`."
  def _sym():
    freq = lambda col,l: col.has.get(k,0) / col.n
    return [(k,k) for k in best.has if freq(best,k) > freq(rest,k)]
  def _num():
    a,z,mu,d = -1E30, 1E30, best.mu, abs(best.mu - cross(best.mu,rest.mu,div(best), div(rest)))
    tmp = [(a, mu),(a, mu+d/2),(a, mu+d)] if mu < rest.mu else [(mu, z),(mu-d/2, z),(mu-d, z)]
    f   = best.prep
    return [ (f(lo), f(hi)) for lo,hi in tmp + [(mu-d, mu+d), (mu-d/2, mu+d/2)] ]
  for lo,hi in set( _sym() if best.this is SYM else _num() ):
    yield KEY(best,lo,hi)

def bore(best,rest):
  def _bore(best1, rest1, b4):
    if step < len(best1.rows):
      new = sorted(rules(best1,rest1,len(best1.rows), len(best2.rows)),reverse=True
      bestrows = [row for row in best.rows if selects(new,row)]
      restrows = [row for row in rest.rows if selects(new,row)]
      if len(best2.rows) !=< len(best1.rows): 
  all = sorted(rules(bests,rests),reverse=True)
  out = [(score,rule) for rule in alll]
  out  = RULE()
  candidates= sorted(rules(best,rest,len(best1.rows), len(best2.rows)),reverse=True)[0]
  _bore(bests,rests, -1)
  return out

#----------------------------------------------------------------------------------------
def coerce(x):
  try: return literal_eval(x)
  except: return x

def csv(file):
  if file=="-": file=None
  with file_or_stdin(file) as src:
    for line in src:
      line = re.sub(r'([\n\t\r"\' ]|#.*)', '', line)
      if line:
        yield [coerce(s.strip()) for s in line.split(",")]
#----------------------------------------------------------------------------------------
class EGS:
  all = locals()

  def Help():
    "Show help"
    print("./tiny.py --ACTION\n\nACTIONS:")
    [print(f"  --{k}\t: {f.__doc__}") for k,f in EGS.all.items() if k[0].isupper()]

  def Rows():
    "Can we read rows from a csv file?"
    [print(row) for row in csv(the.file)]

  def Data():
    "Can we load a csv file into one of our DATAs?"
    print(DATA(csv(the.file)).cols.x[-1])

  def Stats():
    "Can we summarize the distributions in a DATA?"
    print(stats(DATA(csv(the.file))))

  def Better():
    "Can we sort two rows?"
    d=DATA(csv(the.file))
    rows=ordered(d)
    best = rows[:10]
    rest = random.sample(rows[10:],90)
    print("best", stats(clone(d,best)))
    print("rest", stats(clone(d,rest)))

  def Roots():
    print(valleyBetween( 2.5,5,1,1)) # --> 3.75

  def Bins():
    "Can we sort two rows?"
    d=DATA(csv(the.file))
    rows=ordered(d)
    best = clone(d, rows[:20])
    rest = clone(d,random.sample(rows[20:],60))
    for at,name,lo,hi in bin(best,rest):
      print(at,name,lo,hi)

  def Ranges():
    "Can we sort two rows?"
    d= DATA(csv(the.file))
    rows = ordered(d)
    n = int(len(rows)**the.min)
    bests = clone(d, rows[:n])
    rests = clone(d,random.sample(rows[n:],n*the.rest))
    score,at,name,lo,hi = sorted(selects(bests,rests),reverse=True)[0]
    print(score,at,name,lo,hi)

  def Bore():
    "Can we sort two rows?"
    d    = DATA(csv(the.file))
    rows = ordered(d)
    n    = int(len(rows)**the.min)
    print(n, n*the.rest)
    bore( clone(d, rows[:n]),
          clone(d, random.sample(rows[n:], n*the.rest)))
#--------------------------------------------------------------------------------------
if sys.argv[1:]: 
  random.seed(the.seed)
  EGS.all[sys.argv[1][2:]]()

File: tiny.lua
#!/usr/bin/env lua
-- <!--- vim : set et sts=2 sw=2 ts=2 : --->
local l=require"lib"
local the,help = l.settings [[

tiny: multi-goal semi-supervised explanation
(c) 2023 Tim Menzies <timm@ieee.org> BSD-2
  
USAGE: ./tiny.lua [OPTIONS] [-g ACTIONS]
  
OPTIONS:
  -f  --file    data file                          = ../data/auto93.csv
  -g  --go      start-up action                    = nothing
  -G  --Goal    plan or monitor or xplore or doubt = plan
  -h  --help    show help                          = false
  -s  --seed    random number seed                 = 93716221]]

local eg, o, obj, oo, rnd = l.eg, l.o, l.obj, l.oo, l.rnd
local SYM,NUM,COLS,DATA = obj"SYM", obj"NUM", obj"COLS", obj"DATA"
local CONDTION,RULE = obj"CONDITION", obj"RULE"
local COL
-------------------------------------------------------------------------------
-- ### COLS

-- `COL`umns can be `NUM`eric or `SYM`bolic. Upper case names denote `NUM`s.
-- All `COL`s know their name, their column loc`at`ion, their count `n` of items seen.
function COL(n,s) 
  return (s or ""):find"^[A-Z]" and NUM(n,s) or SYM(n,s) end
-------------------------------------------------------------------------------
-- ### SYM

-- `SYM`s uses `has` to count symbols seen so far (and the most common symbol is the `mode`).
function SYM:new(n,s) 
  return {name=s or "", at=n or 0, n=0, has={}, most=0, mode=nil}  end

-- Update a `SYM`bol.
function SYM:add(x)
  if x=="?" then return end
  self.n      = 1 + self.n
  self.has[x] = 1 + (self.has[x] or 0)
  if self.has[x] > self.most then self.most,self.mode = self.has[x],x end end

-- `mid,div` returns central tendency  and diversity.
function SYM:mid(_) return self.mode end
function SYM:div(  nPlaces,     e) 
  e = 0
  for _,v in pairs(self.has) do e = e - v/self.n * math.log(v/self.n,2) end
  return nPlaces and rnd(e,nPlaces) or e end
-------------------------------------------------------------------------------
-- ### NUM

-- `NUM`s tracks the smallest and biggest number seen to far (in `lo` and `hi`) 
-- as well as the the mean `mu` and second moment `m2` (used to find standard deviation).
-- Also, `pretty` controls how we report numbers (and this switches to "%g" 
-- if we ever see a float).  Any name ending with `-` is something we want to _minimize_ 
-- (so we give it a non-positive weight).
function NUM:new(n,s) 
  return {name=s or "",  at=n or 0,  n=0, 
          lo=l.big, hi= -l.big, mu=0,  m2=0, 
          pretty="%.0f", want=(s or ""):find"-$" and 0 or 1} end

-- Update a `NUM`ber.
function NUM:add(n,    d)
  if n=="?" then return end
  self.n  = 1 + self.n
  self.lo = math.min(n, self.lo)
  self.hi = math.max(n, self.hi)
  d    = n - self.mu
  self.mu = self.mu + d/self.n
  self.m2 = self.m2 + d*(n - self.mu)
  if math.type(n) == "float" then self.pretty = "%g" end end

-- `mid,div` returns central tendency  and diversity.
function NUM:mid(  nPlaces) 
  return nPlaces and rnd(self.mu,nPlaces) or self.mu end
function NUM:div(  nPlaces,      sd)
  sd = (self.m2/(self.n - 1))^.5; return nPlaces and rnd(sd,nPlaces) or sd end

-- Return `n` mapped 0..1, min..max.
function NUM:norm(n)
  return n=="?" and x or (n - self.lo)/(self.hi - self.lo + 1/l.big) end

function SYM.keys1(i,j,     t)
  t = {}
  for k,n1 in pairs(i.has) do if n1/i.n < (j.has[k] or 0)/j.n then l.push(t,k) end end
  return t end

% sets keys
% scores
function NUM.keys1(i,j)
  a, z, mu = -1E30, 1E30, best.mu
  d = math.abs(best.mu - cross(best.mu, rest.mu, best:div(), rest:div()))
  tmp = mu < rest.mu and {{a, mu},{a, mu+d/2},{a, mu+d}} or {{mu, z},{mu-d/2, z},{mu-d, z}}
  for _,x in pairs{{mu-d, mu+d}, {mu-d/2, mu+d/2}} do l.push(tmp,x) end
  f = function(n) return l.fmt(best.pretty,n) end 
  return map(tmp,function(x) return {f(x[1]), f(x[2])} end)



local function cross(mu1,mu2,sd1,sd2)
  local sd1,sd2,a,b,c,x1,x2
  if mu2 < mu1 then return cross(mu2,mu1,sd2,sd1) end
  if sd1==0 or sd2==0 or sd1==sd2 then return (mu1+mu2)/2 end
  a  = 1/(2*sd1^2) - 1/(2*sd2^2)
  b  = mu2/(sd2^2) - mu1/(sd1^2)
  c  = mu1^2 /(2*sd1^2) - mu2^2 / (2*sd2^2) - math.log(sd2/sd1)
  x1 = (-b + (b^2 - 4*a*c)^.5)/(2*a)
  x2 = (-b - (b^2 - 4*a*c)^.5)/(2*a)
  return mu1 <= x1 <= mu2 and x1 or x2 end
-------------------------------------------------------------------------------
-- ### COLS

-- `COLS` convert  list of column names (as strings) into `SYM`s or `NUM`s. 
-- Column names ending in a goal symbol (`+-!`) are dependent `y` features 
-- (and the others are independent `x` features). 
function COLS:new(ss) 
  self.x, self.y, self.all, self.names = {},{},{},ss
  for n,s in pairs(ss) do
    local col = l.push(self.all, COL(n,s))
    if not s:find"X$" then 
      l.push(s:find"[!+-]$" and self.y or self.x, col)
      if s:find"!$" then self.klass=col end  end end  end

-- Update a `COL` with info from `t`.
function COLS:add(t)
  for _,cols in pairs{self.x, self.y} do
    for _,col in pairs(cols) do 
      col:add(t[col.at]) end end 
  return t end
-------------------------------------------------------------------------------
-- ### DATA
function DATA:new(  ts)
  self.rows,self.cols = {},nil
  if   type(ts)=="string" 
  then l.csv(ts, function(t) self:add(t) end)  
  else for _,t in pairs(ts or {}) do self:add(t) end end end

function DATA:add(t)
  if self.cols then l.push(self.rows, self.cols:add(t)) else self.cols=COLS(t) end end

function DATA:clone(  ts,     data1)
  data1 = DATA()
  data1.cols = COLS(self.cols.names)
  for _,t in pairs(ts or {}) do data1:add(t) end
  return data1 end

function DATA:d2h(t,     d)
  d = 0
  for _,col in pairs(self.cols.y) do d = d + (col.want - col:norm(t[col.at]))^2 end
  return d^.5 end

function DATA:sort()
  return l.sort(self.rows, function(t1,t2) return self:d2h(t1) < self:d2h(t2) end) end

function DATA:stats(  what,cols,nPlaces,     t)
  t = {N=#self.rows}
  for _,col in pairs(cols or self.cols.y) do
    t[col.name] =rnd(getmetatable(col)[what or "mid"](col),nPlaces) end
  return t end
-------------------------------------------------------------------------------
function TEST:new(col,lo,hi)
  return {at=col.at, b=0, r=0,lo=lo, hi=hi, name=col.name} end

function TEST:covers(t)
  local x = t[self.at]
  if x=="?" or self.lo <= x and x <= self.hi then return t end end 

function RULE:new(tests)
  self.cols, self.b, self.r = {}, 0,0,0,0
  for _,test is pairs(tests or {}) do self:add(test) end end

  -- disjunct, spread the area XXX
function RULE:add(test,      olds,skip)
  olds = self.cols[new.at] = self.cols[new.at] or {}
  skip = false
  for _,old in pairs(olds) do
    local a0, z0, a1, z1 = old.lo, old.hi, test.lo, test.hi
    if z1 >= a0 and z1 <= z0 and a0 <= a1 then old.hi = test.hi; skip = true end 
    if a1 >= a0 and a1 <= z0 and z1 >= z0 then old.lo = test.lo; skip = true end end
  if not skip then l.push(olds,new) end end

function RULE:covers(t,    _ors)
  function _ors(tests)
    for _,test in pairs(tests) do 
      if test:covers(t) then return true end end end
  for _,tests in pairs(self.cols) do
    if not _ors(tests) then return nil end end
  return t end

function br(thing,bests,rests)
  local B, R = #bests.rows, #rests.rows 
  bs  = map(bests.rows, function(t) return thing:covers(t) end)
  rs  = map(rests.rows, function(t) return thing:covers(t) end)
  thing.b, thing.r = #bs/(B + 1/l.big),  #rs/(R + 1/l.big) end

function goal(thing)
  b, r = thing.b, thing.r
  if the.Goal == "plan"    then return b^2/(b+r) end
  if the.Goal == "monitor" then return r^2/(b+r) end
  if the.Goal == "xplore"  then return 1/(b+r)   end
  if the.Goal == "doubt"   then return (b+r)/math.abs(b-r) end end

-------------------------------------------------------------------------------
-- ## Demos
eg("the",  function() oo(the) end)
eg("rnd",  function() return 3.14 == rnd(math.pi) end)
eg("rand", function(     num,d) 
  num = NUM()
  for _ = 1,1000 do num:add(l.rand()^2) end
  d = num:div()
  return .3 < d and d < .31  end)

eg("ent", function(     sym,e,s) 
  sym = SYM()
  s="aaaabbc"
  for c in s:gmatch"." do sym:add(c) end
  e = sym:div()
  return 1.37 < e and e < 1.38 end)

eg("cross", function()
  return 3.75 == cross(2.5,5,1,1) end)

eg("cols", function()
  l.map(COLS({"Name", "Age", "Mpg-", "room"}).all, print) end)

eg("csv", function()
  l.csv(the.file,oo) end)

eg("data", function(       data)
  data = DATA(the.file) 
  oo(data:stats()) end)

eg("slice", function(      t)
  t = {10,20,30,40,50,60,70,80,90,100}
  print(-2,"", o(l.slice(t,-2))) 
  print( 1, 2, o(l.slice(t,1,2))) 
  print( 2,"", o(l.slice(t,2)))  end)

eg("sort", function(       data,ts1,ts2,ts3)
  data = DATA(the.file) 
  ts1 = data:sort() 
  ts2 = l.slice(ts1,1,30)
  ts3 = l.slice(ts1,-120)
  print("all ", o(data:stats()))
  print("best", o(data:clone(ts2):stats()))
  print("rest", o(data:clone(ts3):stats()))
end)
-------------------------------------------------------------------------------
-- ## Start-up
if   not pcall(debug.getlocal,4,1) 
then the=l.cli(the, help); l.run(the) end

return {COL=COL, COLS=COLS, DATA=DATA, NUM=NUM, SYM=SYM} 
